src/store/conversation/conversationSlice.ts

```diff
@@ -0,0 +1,217 @@
+import { createSlice, createEntityAdapter } from '@reduxjs/toolkit';
+import { Conversation } from '@/types/Conversation';
+import { conversationActions } from './conversationActions';
+import { findPendingMessageIndex } from '@/utils/conversationUtils';
+
+import { MESSAGE_TYPES } from '@/constants';
+import { Message } from '@/types/Message';
+import { PendingMessage } from './conversationTypes';
+
+export interface ConversationState {
+  meta: {
+    mineCount: number;
+    unassignedCount: number;
+    allCount: number;
+  };
+  error: string | null;
+  uiFlags: {
+    isLoadingConversations: boolean;
+    isLoadingMessages: boolean;
+    isAllConversationsFetched: boolean;
+    isAllMessagesFetched: boolean;
+    isConversationFetching: boolean;
+    isChangingConversationStatus: boolean;
+  };
+}
+
+export const conversationAdapter = createEntityAdapter<Conversation>({
+  selectId: conversation => conversation.id,
+});
+
+const initialState = conversationAdapter.getInitialState<ConversationState>({
+  meta: {
+    mineCount: 0,
+    unassignedCount: 0,
+    allCount: 0,
+  },
+  error: null,
+  uiFlags: {
+    isLoadingConversations: false,
+    isAllConversationsFetched: false,
+    isLoadingMessages: false,
+    isAllMessagesFetched: false,
+    isConversationFetching: false,
+    isChangingConversationStatus: false,
+  },
+});
+
+const conversationSlice = createSlice({
+  name: 'conversation',
+  initialState,
+  reducers: {
+    clearAllConversations: conversationAdapter.removeAll,
+    addConversation: (state, action) => {
+      const conversation = action.payload;
+      conversationAdapter.addOne(state, conversation);
+    },
+    updateConversation: (state, action) => {
+      const conversation = action.payload as Conversation;
+      const conversationIds = conversationAdapter.getSelectors().selectIds(state);
+      if (conversationIds.includes(conversation.id)) {
+        const { messages, ...conversationAttributes } = conversation;
+        conversationAdapter.updateOne(state, {
+          id: conversation.id,
+          changes: conversationAttributes,
+        });
+      } else {
+        conversationAdapter.addOne(state, conversation);
+      }
+    },
+    addOrUpdateMessage: (state, action) => {
+      const message = action.payload as PendingMessage | Message;
+
+      const { conversationId } = message;
+      if (!conversationId) {
+        return;
+      }
+
+      const conversation = state.entities[conversationId];
```

Comment:
This is a mistake we made in the dashboard. Let us try to avoid it here. Move the messages to a message store

{ messageId: messageObject } format. To get the messages in order, we could get all message by conversationId. Adding a new message or updating message becomes easier this way.